core/dom/range.js

```diff
@@ -2843,46 +2843,231 @@ CKEDITOR.dom.range = function( root ) {
 		 */
 		getClientRects: ( function() {
 			if ( this.document.getSelection !== undefined ) {
-				return function() {
+				return function( isAbsolute ) {
 					// We need to create native range so we can call native getClientRects.
-					var range = document.createRange(),
-						rectArray;
+					var range = this.root.getDocument().$.createRange(),
+						rectList;
 
 					range.setStart( this.startContainer.$, this.startOffset );
 					range.setEnd( this.endContainer.$, this.endOffset );
 
-					rectArray = CKEDITOR.tools.array.map( range.getClientRects(), function( item ) {
-						return convertRect( item );
-					} );
+					rectList = range.getClientRects();
 
-					// In some cases ( eg. ranges contain only image ) IE will return empty rectList.
-					if ( !range.collapsed && !rectArray.length ) {
-						rectArray = [ getRect( this.createBookmark() ) ];
+					rectList = fixWidgetsRects( rectList, this );
+
+					if ( !rectList.length ) {
+						rectList = fixEmptyRectList( rectList, range, this );
 					}
 
 					range.detach();
 
-					return rectArray;
+					return CKEDITOR.tools.array.map( rectList, function( item ) {
+						return convertRect( item, isAbsolute, this );
+					}, this );
 				};
 			} else {
-				return function() {
-					return [ getRect( this.createBookmark() ) ];
+				return function( isAbsolute ) {
+					return [ convertRect( getRect( this.createBookmark() ), isAbsolute, this ) ];
 				};
 			}
 
+			// Remove all widget rects except for outermost one.
+			function fixWidgetsRects( rectList, context ) {
+				var walker = new CKEDITOR.dom.walker( context ),
+					rectArray = CKEDITOR.tools.array.map( rectList, function( item ) {
+						return item;
+					} ),
+					newRange = new CKEDITOR.dom.range( context.root ),
+					widgetElements = [],
+					widgetIds = [],
+					widgetRects,
+					widgetRange,
+					nodeList,
+					documentFragment,
+					moveStart,
+					moveEnd;
+
+				// In case of ranges start and end container set as widget wrapper, document container won't contain wrapper and we cant find its id.
+				// Let's move ranges to parent element to fix that.
+				if ( context.startContainer instanceof CKEDITOR.dom.element ) {
+					moveStart = context.startOffset === 0 && context.startContainer.hasAttribute( 'data-widget' );
+				}
+				if ( context.endContainer instanceof CKEDITOR.dom.element ) {
+					moveEnd = context.endOffset === ( context.endContainer.getChildCount ? context.endContainer.getChildCount() : context.endContainer.length );
+					moveEnd = moveEnd && context.endContainer.hasAttribute( 'data-widget' );
+				}
+
+				if ( moveStart ) {
+					newRange.setStart( context.startContainer.getParent(), context.startContainer.getIndex() );
+				}
+				if ( moveEnd ) {
+					newRange.setEnd( context.endContainer.getParent(), context.endContainer.getIndex() + 1 );
+				}
+				if ( moveStart || moveEnd ) {
+					context = newRange;
+				}
+
+				documentFragment = context.cloneContents();
```

Comment:
Can't this logic be greatly simplified by omitting walker altogether? Something like:

```javascript
var selectedNodes = context.cloneContents(),
	widgets = CKEDITOR.dom.document.prototype.find.call( selectedNodes, '[data-cke-widget-id]' ).toArray();

CKEDITOR.tools.forEach( widgets, function() {
	// Logic for every widget.
} );
```